From 69f357fec81759cec5a57d1a1d934ab5445654f8 Mon Sep 17 00:00:00 2001
From: t <tristan75015@gmail.com>
Date: Thu, 8 Oct 2026 06:03:07 +0200
Subject: [PATCH] Integration of a batch for the behaviours generated by
 generic-parallel

- integrate(BatchExecutionPolicy, MaterialDataManager, options, dt[, b, e])
  integrates the points of a material data manager as a batch. The fields
  of the batch refer to the arrays of the manager, the material properties
  and the external state variables being given variable by variable
  (uniform values or values stored point by point): no data is copied.
  Invalid arguments and errors of the backend are reported by exceptions,
  never as failures of the integration. In case of failure, the index of
  the first point where the integration failed is reported.
- The integration functions of a material data manager switch to the
  integration of a batch for those behaviours, with the execution policy
  of the manager (batch_execution_policy). integrate_debug is refused.
- The uniform values stored by the manager are in the memory of the host:
  they are refused for the cuda and hip backends, which must be given
  spans on memory accessible by the device.
- The libraries are never unloaded (RTLD_NODELETE): the threads of TBB,
  used by the parallel algorithms of the C++ standard library, were still
  alive when the libraries were closed at the end of the process
  (intermittent crash at exit).

GenericParallelIntegrateTest compares the integration of batches with the
integration point by point of the generic interface (Norton over several
time steps, plasticity with the temperature given point by point, elastic
behaviour with material properties given point by point and uniform,
ranges of points, prediction, failure at all points).
---
 .../MGIS/Behaviour/BatchExecutionPolicy.hxx   |  42 +++
 include/MGIS/Behaviour/Integrate.hxx          |  50 +++
 .../MGIS/Behaviour/MaterialDataManager.hxx    |   6 +
 src/Integrate.cxx                             | 285 ++++++++++++++++++
 src/LibrariesManager.cxx                      |   6 +-
 tests/CMakeLists.txt                          |  15 +
 tests/GenericParallelIntegrateTest.cxx        | 265 ++++++++++++++++
 7 files changed, 668 insertions(+), 1 deletion(-)
 create mode 100644 include/MGIS/Behaviour/BatchExecutionPolicy.hxx
 create mode 100644 tests/GenericParallelIntegrateTest.cxx

diff --git a/include/MGIS/Behaviour/BatchExecutionPolicy.hxx b/include/MGIS/Behaviour/BatchExecutionPolicy.hxx
new file mode 100644
index 0000000..d34c460
--- /dev/null
+++ b/include/MGIS/Behaviour/BatchExecutionPolicy.hxx
@@ -0,0 +1,42 @@
+/*!
+ * \file   include/MGIS/Behaviour/BatchExecutionPolicy.hxx
+ * \brief  This file declares the `BatchExecutionPolicy` structure.
+ * \date   09/10/2026
+ * \copyright Use, modification and distribution are subject
+ * to one of the following licences:
+ * - GNU Lesser General Public License (LGPL), Version 3.0. (See accompanying
+ *   file LGPL-3.0.txt)
+ * - CECILL-C,  Version 1.0 (See accompanying files
+ *   CeCILL-C_V1-en.txt and CeCILL-C_V1-fr.txt).
+ */
+
+#ifndef LIB_MGIS_BEHAVIOUR_BATCHEXECUTIONPOLICY_HXX
+#define LIB_MGIS_BEHAVIOUR_BATCHEXECUTIONPOLICY_HXX
+
+namespace mgis::behaviour {
+
+  /*!
+   * \brief execution of the computations on a batch of integration points,
+   * for the behaviours generated by the `generic-parallel` interface of
+   * `MFront`. The location of the data is a property of the data, not of
+   * the execution.
+   */
+  struct BatchExecutionPolicy {
+    /*!
+     * \brief queue on which the computations are submitted. Its meaning
+     * depends on the backend of the behaviour (`cudaStream_t` or
+     * `hipStream_t`, pointer to a `sycl::queue` or to an instance of a
+     * Kokkos execution space). A null pointer denotes the default queue, if
+     * meaningful for the backend.
+     */
+    void* queue = nullptr;
+    /*!
+     * \brief if true, the stack size of the threads of the device is never
+     * modified and an error is reported if it is too small
+     */
+    bool preserve_stack_size = false;
+  };
+
+}  // end of namespace mgis::behaviour
+
+#endif /* LIB_MGIS_BEHAVIOUR_BATCHEXECUTIONPOLICY_HXX */
diff --git a/include/MGIS/Behaviour/Integrate.hxx b/include/MGIS/Behaviour/Integrate.hxx
index 8cc099d..6bd2b75 100644
--- a/include/MGIS/Behaviour/Integrate.hxx
+++ b/include/MGIS/Behaviour/Integrate.hxx
@@ -20,6 +20,7 @@
 #include <vector>
 #include "MGIS/Config.hxx"
 #include "MGIS/Behaviour/BehaviourDataView.hxx"
+#include "MGIS/Behaviour/BatchExecutionPolicy.hxx"
 
 namespace mgis {
 
@@ -323,9 +324,58 @@ namespace mgis::behaviour {
    *
    * \note if required, the memory associated with the tangent operator blocks
    * is automatically allocated.
+   * \note for a behaviour generated by the `generic-parallel` interface, the
+   * points are integrated as a batch with the execution policy of the
+   * material data manager (see the overload taking an execution policy).
    */
   MGIS_EXPORT BehaviourIntegrationResult integrate(
       MaterialDataManager&, const BehaviourIntegrationOptions&, const real);
+  /*!
+   * \brief integrate a behaviour generated by the `generic-parallel`
+   * interface on all the points of a material data manager, as a batch.
+   *
+   * The fields of the batch refer to the arrays of the material data
+   * manager: no data is copied. Those arrays must be accessible by the device
+   * on which the computations are performed (see the backend of the
+   * behaviour).
+   *
+   * Contrary to the integration point by point, all the points are
+   * integrated, even if the integration fails at some points. In case of
+   * failure, the index of the first point where the integration failed is
+   * reported (member `n` of the result). Invalid arguments and errors of the
+   * backend are reported by exceptions: they are never reported as a failure
+   * of the integration.
+   *
+   * \return the result of the behaviour integration.
+   * \param[in] p: execution policy
+   * \param[in,out] m: material data manager
+   * \param[in] opts: description of the operation to be performed
+   * \param[in] dt: time step
+   */
+  MGIS_EXPORT BehaviourIntegrationResult
+  integrate(const BatchExecutionPolicy&,
+            MaterialDataManager&,
+            const BehaviourIntegrationOptions&,
+            const real);
+  /*!
+   * \brief integrate a behaviour generated by the `generic-parallel`
+   * interface on a range of integration points, as a batch (see the previous
+   * overload).
+   * \return the result of the behaviour integration.
+   * \param[in] p: execution policy
+   * \param[in,out] m: material data manager
+   * \param[in] opts: description of the operation to be performed
+   * \param[in] dt: time step
+   * \param[in] b: first index of the range
+   * \param[in] e: last index of the range
+   */
+  MGIS_EXPORT BehaviourIntegrationResult
+  integrate(const BatchExecutionPolicy&,
+            MaterialDataManager&,
+            const BehaviourIntegrationOptions&,
+            const real,
+            const size_type,
+            const size_type);
   //! \brief debugging version
   MGIS_EXPORT BehaviourIntegrationResult integrate_debug(
       MaterialDataManager&, const BehaviourIntegrationOptions&, const real);
diff --git a/include/MGIS/Behaviour/MaterialDataManager.hxx b/include/MGIS/Behaviour/MaterialDataManager.hxx
index 7073cc4..3238aca 100644
--- a/include/MGIS/Behaviour/MaterialDataManager.hxx
+++ b/include/MGIS/Behaviour/MaterialDataManager.hxx
@@ -25,6 +25,7 @@
 
 #include "MGIS/Config.hxx"
 #include "MGIS/Behaviour/MaterialStateManager.hxx"
+#include "MGIS/Behaviour/BatchExecutionPolicy.hxx"
 
 namespace mgis::behaviour {
 
@@ -232,6 +233,11 @@ namespace mgis::behaviour {
      * behaviour to request an increase of the time step.
      */
     real rdt = 1;
+    /*!
+     * \brief execution policy used by the functions which do not take one,
+     * for the behaviours generated by the `generic-parallel` interface
+     */
+    BatchExecutionPolicy batch_execution_policy;
     //! \brief view on the speed of sound.
     std::span<real> speed_of_sound;
     //! \brief number of integration points
diff --git a/src/Integrate.cxx b/src/Integrate.cxx
index 40f5d16..d799dd2 100644
--- a/src/Integrate.cxx
+++ b/src/Integrate.cxx
@@ -13,6 +13,9 @@
 
 #include <map>
 #include <tuple>
+#include <variant>
+#include <string_view>
+#include <type_traits>
 #include <thread>
 #include <memory>
 #include <cstdlib>
@@ -21,6 +24,7 @@
 #include "MGIS/Raise.hxx"
 #include "MGIS/ThreadPool.hxx"
 #include "MGIS/Behaviour/MaterialDataManager.hxx"
+#include "MGIS/Behaviour/BatchDataView.hxx"
 #include "MGIS/Behaviour/BehaviourIntegrationFailureAnalyser.hxx"
 #include "MGIS/Behaviour/Integrate.hxx"
 
@@ -300,6 +304,243 @@ namespace mgis::behaviour::internals {
     }
   }  // end of checkIntegrationPointsRange
 
+  /*!
+   * \return true if the backend of a behaviour generated by the
+   * `generic-parallel` interface executes the computations on a device which
+   * can't access the memory of the host
+   * \param[in] b: behaviour
+   */
+  static bool isDeviceOnlyBackend(const Behaviour& b) noexcept {
+    return (b.backend == "cuda") || (b.backend == "hip");
+  }  // end of isDeviceOnlyBackend
+
+  /*!
+   * \return the field of a uniform value stored by the material data
+   * manager, in the memory of the host
+   * \param[in] m: material data manager
+   * \param[in] n: name of the variable
+   * \param[in] v: value
+   */
+  static ConstField makeUniformBatchField(const MaterialDataManager& m,
+                                          const std::string_view n,
+                                          const real& v) {
+    if (isDeviceOnlyBackend(m.b)) {
+      mgis::raise("makeUniformBatchField: the uniform value of '" +
+                  std::string{n} +
+                  "' is stored in the memory of the host, which is not "
+                  "accessible by the device of the backend '" +
+                  m.b.backend +
+                  "': this value must be given as a span on memory "
+                  "accessible by the device");
+    }
+    return {&v, 0, 1};
+  }  // end of makeUniformBatchField
+
+  /*!
+   * \return the field of a material property or of an external state
+   * variable on the points of a batch, starting at the point `b`. The values
+   * are either uniform or stored point by point.
+   * \param[in] m: material data manager
+   * \param[in] v: variable
+   * \param[in] fh: field holder
+   * \param[in] b: first point of the batch
+   */
+  static ConstField makeBatchField(const MaterialDataManager& m,
+                                   const Variable& v,
+                                   const MaterialStateManager::FieldHolder& fh,
+                                   const size_type b) {
+    if (v.type == Variable::ARRAY) {
+      mgis::raise("makeBatchField: the arrays of variables ('" + v.name +
+                  "') are not supported by the integration of a batch");
+    }
+    const auto s = getVariableSize(v, m.b.hypothesis);
+    const auto from_values = [&m, &v, s, b](const auto& values) -> ConstField {
+      if (values.size() == s) {
+        return {values.data(), 0, 1};
+      }
+      if (values.size() != m.n * s) {
+        mgis::raise("makeBatchField: invalid number of values for '" + v.name +
+                    "' (" + std::to_string(values.size()) + " given)");
+      }
+      return {values.data() + b * s, s, 1};
+    };
+    if (std::holds_alternative<std::monostate>(fh)) {
+      mgis::raise("makeBatchField: uninitialized field holder '" + v.name +
+                  "'");
+    }
+    if (std::holds_alternative<std::span<const real>>(fh)) {
+      return from_values(std::get<std::span<const real>>(fh));
+    }
+    const auto& mfh = std::get<MaterialStateManager::MutableFieldHolder>(fh);
+    if (std::holds_alternative<real>(mfh.value)) {
+      if (s != 1) {
+        mgis::raise("makeBatchField: invalid uniform value for '" + v.name +
+                    "'");
+      }
+      return makeUniformBatchField(m, v.name, std::get<real>(mfh.value));
+    }
+    if (std::holds_alternative<std::span<real>>(mfh.value)) {
+      return from_values(std::get<std::span<real>>(mfh.value));
+    }
+    return from_values(std::get<std::vector<real>>(mfh.value));
+  }  // end of makeBatchField
+
+  /*!
+   * \return the fields of a group of variables (material properties or
+   * external state variables), variable by variable
+   * \param[in] m: material data manager
+   * \param[in] variables: variables
+   * \param[in] values: field holders
+   * \param[in] b: first point of the batch
+   */
+  static std::vector<ConstField> makeBatchFields(
+      const MaterialDataManager& m,
+      const std::vector<Variable>& variables,
+      const std::map<std::string,
+                     MaterialStateManager::FieldHolder,
+                     std::less<>>& values,
+      const size_type b) {
+    auto fields = std::vector<ConstField>{};
+    fields.reserve(variables.size());
+    for (const auto& v : variables) {
+      const auto p = values.find(v.name);
+      if (p == values.end()) {
+        mgis::raise("makeBatchFields: no variable named '" + v.name +
+                    "' declared");
+      }
+      fields.push_back(makeBatchField(m, v, p->second, b));
+    }
+    return fields;
+  }  // end of makeBatchFields
+
+  //! \return the field of the mass density of a state
+  static ConstField makeMassDensityBatchField(const MaterialDataManager& m,
+                                              const MaterialStateManager& s,
+                                              const size_type b) {
+    if (!s.mass_density.has_value()) {
+      return {nullptr, 0, 0};
+    }
+    const auto& v = s.mass_density->value;
+    if (std::holds_alternative<real>(v)) {
+      return makeUniformBatchField(m, "mass density", std::get<real>(v));
+    }
+    const auto* const values = std::holds_alternative<std::span<real>>(v)
+                                   ? std::get<std::span<real>>(v).data()
+                                   : std::get<std::vector<real>>(v).data();
+    return {values + b, 1, 1};
+  }  // end of makeMassDensityBatchField
+
+  //! \return a field holding the values of an array, point by point
+  template <typename ValueType>
+  static auto makeArrayBatchField(const std::span<ValueType> values,
+                                  const size_type stride,
+                                  const size_type b) {
+    using Field = std::conditional_t<std::is_const_v<ValueType>, ConstField,
+                                     mgis::behaviour::Field>;
+    if (values.empty()) {
+      return Field{nullptr, 0, 0};
+    }
+    return Field{values.data() + b * stride, stride, 1};
+  }  // end of makeBatchField
+
+  /*!
+   * \brief integrate a behaviour generated by the `generic-parallel`
+   * interface on the points `[b, e)` of a material data manager.
+   */
+  static BehaviourIntegrationResult integrateBatch(
+      const BatchExecutionPolicy& p,
+      MaterialDataManager& m,
+      const BehaviourIntegrationOptions& opts,
+      const real dt,
+      const size_type b,
+      const size_type e) {
+    if (m.b.batch_b == nullptr) {
+      mgis::raise(
+          "integrate: the behaviour has not been generated by the "
+          "generic-parallel interface");
+    }
+    auto r = BehaviourIntegrationResult{};
+    if (b == e) {
+      return r;
+    }
+    // fields of the material properties and of the external state
+    // variables, variable by variable
+    const auto mps0 = makeBatchFields(m, m.b.mps, m.s0.material_properties, b);
+    const auto mps1 = makeBatchFields(m, m.b.mps, m.s1.material_properties, b);
+    const auto esvs0 =
+        makeBatchFields(m, m.b.esvs, m.s0.external_state_variables, b);
+    const auto esvs1 =
+        makeBatchFields(m, m.b.esvs, m.s1.external_state_variables, b);
+    auto d = BatchDataView{};
+    d.number_of_points = e - b;
+    d.dt = dt;
+    d.options[0] = encodeBehaviourIntegrationOptions(opts);
+    for (size_type i = 0; (i != m.b.options.size()) && (i != 2); ++i) {
+      d.options[i + 1] = m.b.options[i];
+    }
+    // same initial value of the time step scaling factor as the integration
+    // point by point
+    d.rdt = r.time_step_increase_factor;
+    if ((opts.integration_type !=
+         IntegrationType::INTEGRATION_NO_TANGENT_OPERATOR) &&
+        (m.K_stride != 0)) {
+      d.tangent_operators = makeArrayBatchField(m.K, m.K_stride, b);
+    }
+    if (opts.compute_speed_of_sound) {
+      d.speed_of_sound = makeArrayBatchField(m.speed_of_sound, 1, b);
+    }
+    // state at the beginning of the time step
+    const auto& s0 = m.s0;
+    d.s0.gradients = makeArrayBatchField(std::span<const real>(s0.gradients),
+                                         s0.gradients_stride, b);
+    d.s0.thermodynamic_forces =
+        makeArrayBatchField(std::span<const real>(s0.thermodynamic_forces),
+                            s0.thermodynamic_forces_stride, b);
+    d.s0.mass_density = makeMassDensityBatchField(m, s0, b);
+    d.s0.internal_state_variables =
+        makeArrayBatchField(std::span<const real>(s0.internal_state_variables),
+                            s0.internal_state_variables_stride, b);
+    d.s0.stored_energy =
+        makeArrayBatchField(std::span<const real>(s0.stored_energies), 1, b);
+    d.s0.dissipated_energy = makeArrayBatchField(
+        std::span<const real>(s0.dissipated_energies), 1, b);
+    d.s0.material_properties_by_variable = mps0.data();
+    d.s0.external_state_variables_by_variable = esvs0.data();
+    // state at the end of the time step
+    auto& s1 = m.s1;
+    d.s1.gradients = makeArrayBatchField(std::span<const real>(s1.gradients),
+                                         s1.gradients_stride, b);
+    d.s1.thermodynamic_forces = makeArrayBatchField(
+        s1.thermodynamic_forces, s1.thermodynamic_forces_stride, b);
+    d.s1.mass_density = makeMassDensityBatchField(m, s1, b);
+    d.s1.internal_state_variables = makeArrayBatchField(
+        s1.internal_state_variables, s1.internal_state_variables_stride, b);
+    d.s1.stored_energy = makeArrayBatchField(s1.stored_energies, 1, b);
+    d.s1.dissipated_energy = makeArrayBatchField(s1.dissipated_energies, 1, b);
+    d.s1.material_properties_by_variable = mps1.data();
+    d.s1.external_state_variables_by_variable = esvs1.data();
+    // synchronous execution
+    d.execution.queue = p.queue;
+    d.execution.statistics = nullptr;
+    d.execution.preserve_stack_size = p.preserve_stack_size ? 1 : 0;
+    auto o = BatchOutputStatus{};
+    const auto status = m.b.batch_b(&o, &d);
+    if ((status == -2) || (status == -3)) {
+      // invalid arguments or error of the backend, which must not be taken
+      // for a failure of the integration
+      o.msg[sizeof(o.msg) - 1] = '\0';
+      mgis::raise("integrate: " + std::string{o.msg});
+    }
+    r.exit_status = status;
+    r.time_step_increase_factor = o.rdt;
+    if (status == -1) {
+      r.n = b + o.first_failure_index;
+      o.msg[sizeof(o.msg) - 1] = '\0';
+      r.error_message = std::string{o.msg};
+    }
+    return r;
+  }  // end of integrateBatch
+
   /*!
    * \brief execute the given initialize function over a range of integration
    * points.
@@ -375,6 +616,11 @@ namespace mgis::behaviour::internals {
       const real dt,
       const size_type b,
       const size_type e) {
+    if (m.b.b == nullptr) {
+      mgis::raise(
+          "integrate: no integration function of a point (the behaviour "
+          "has been generated by the generic-parallel interface)");
+    }
     // workspace
     auto& ws = m.getBehaviourIntegrationWorkSpace();
     auto v = internals::initializeBehaviourDataView(ws);
@@ -425,6 +671,12 @@ namespace mgis::behaviour::internals {
       const size_type b,
       const size_type e,
       const debug::BehaviourIntegrationFailureAnalyser& analyser) {
+    if (m.b.b == nullptr) {
+      mgis::raise(
+          "integrate_debug: not available for the behaviours generated by "
+          "the generic-parallel interface (the generic interface of the same "
+          "behaviour may be used to analyse a failure)");
+    }
     // workspace
     auto& ws = m.getBehaviourIntegrationWorkSpace();
     auto v = internals::initializeBehaviourDataView(ws);
@@ -718,10 +970,32 @@ namespace mgis::behaviour {
   BehaviourIntegrationResult integrate(MaterialDataManager& m,
                                        const BehaviourIntegrationOptions& opts,
                                        const real dt) {
+    if (isGenericParallelBehaviour(m.b)) {
+      return integrate(m.batch_execution_policy, m, opts, dt);
+    }
     internals::allocate(m, opts);
     return internals::integrate(m, opts, dt, 0, m.n);
   }  // end of integrate
 
+  BehaviourIntegrationResult integrate(const BatchExecutionPolicy& p,
+                                       MaterialDataManager& m,
+                                       const BehaviourIntegrationOptions& opts,
+                                       const real dt) {
+    internals::allocate(m, opts);
+    return internals::integrateBatch(p, m, opts, dt, 0, m.n);
+  }  // end of integrate
+
+  BehaviourIntegrationResult integrate(const BatchExecutionPolicy& p,
+                                       MaterialDataManager& m,
+                                       const BehaviourIntegrationOptions& opts,
+                                       const real dt,
+                                       const size_type b,
+                                       const size_type e) {
+    internals::allocate(m, opts);
+    internals::checkIntegrationPointsRange(m, b, e);
+    return internals::integrateBatch(p, m, opts, dt, b, e);
+  }  // end of integrate
+
   BehaviourIntegrationResult integrate_debug(
       MaterialDataManager& m,
       const BehaviourIntegrationOptions& opts,
@@ -744,6 +1018,9 @@ namespace mgis::behaviour {
                                        const real dt,
                                        const size_type b,
                                        const size_type e) {
+    if (isGenericParallelBehaviour(m.b)) {
+      return integrate(m.batch_execution_policy, m, opts, dt, b, e);
+    }
     internals::allocate(m, opts);
     internals::checkIntegrationPointsRange(m, b, e);
     return internals::integrate(m, opts, dt, b, e);
@@ -776,6 +1053,14 @@ namespace mgis::behaviour {
       MaterialDataManager& m,
       const BehaviourIntegrationOptions& opts,
       const real dt) {
+    if (isGenericParallelBehaviour(m.b)) {
+      // the parallelism is the one of the backend of the behaviour
+      auto res = MultiThreadedBehaviourIntegrationResult{};
+      const auto r = integrate(m.batch_execution_policy, m, opts, dt);
+      res.exit_status = r.exit_status;
+      res.results.push_back(r);
+      return res;
+    }
     m.setThreadSafe(true);
     internals::allocate(m, opts);
     // get number of threads
diff --git a/src/LibrariesManager.cxx b/src/LibrariesManager.cxx
index 1606ecd..5aa8b07 100644
--- a/src/LibrariesManager.cxx
+++ b/src/LibrariesManager.cxx
@@ -64,7 +64,11 @@ namespace mgis {
 #if (defined _WIN32 || defined _WIN64) && (!defined __CYGWIN__)
     return ::LoadLibrary(TEXT(l.c_str()));
 #else
-    return ::dlopen(l.c_str(), RTLD_NOW);
+    // the libraries are never unloaded: a library may use threads which are
+    // still alive when the libraries are closed at the end of the process,
+    // for instance the threads of TBB, used by the parallel algorithms of
+    // the C++ standard library (generic-parallel interface)
+    return ::dlopen(l.c_str(), RTLD_NOW | RTLD_NODELETE);
 #endif /* (defined _WIN32 || defined _WIN64) && (!defined __CYGWIN__) */
   }    // end of load_library
 
diff --git a/tests/CMakeLists.txt b/tests/CMakeLists.txt
index 9f83a50..d2e0fe1 100644
--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -140,6 +140,9 @@ if(MGIS_HAVE_MFRONT_SUPPORT)
   # host by the C++ parallel algorithms
   mfront_generic_parallel_behaviours_check_library(
     GenericParallelBehaviourTest stlpar
+    Elasticity
+    Norton
+    NortonFailure
     Plasticity
     OrthotropicElasticity
     InitializeFunctionTest
@@ -158,6 +161,18 @@ if(MGIS_HAVE_MFRONT_SUPPORT)
   set_property(TEST GenericParallelLoadTest
     PROPERTY DEPENDS GenericParallelBehaviourTest BehaviourTest)
   set_test_environment(GenericParallelLoadTest)
+  add_executable(GenericParallelIntegrateTest EXCLUDE_FROM_ALL
+    GenericParallelIntegrateTest.cxx)
+  target_link_libraries(GenericParallelIntegrateTest
+    PRIVATE MFrontGenericInterface)
+  add_test(NAME GenericParallelIntegrateTest
+    COMMAND GenericParallelIntegrateTest
+    "$<TARGET_FILE:GenericParallelBehaviourTest>"
+    "$<TARGET_FILE:BehaviourTest>")
+  add_dependencies(build-tests GenericParallelIntegrateTest)
+  set_property(TEST GenericParallelIntegrateTest
+    PROPERTY DEPENDS GenericParallelBehaviourTest BehaviourTest)
+  set_test_environment(GenericParallelIntegrateTest)
 
   # Tests that depend on behaviour libraries
   add_library_test_with_tfel_version(LoadMaterialPropertyTest LoadMaterialPropertyTest.cxx MaterialPropertyTest)
diff --git a/tests/GenericParallelIntegrateTest.cxx b/tests/GenericParallelIntegrateTest.cxx
new file mode 100644
index 0000000..7ca7333
--- /dev/null
+++ b/tests/GenericParallelIntegrateTest.cxx
@@ -0,0 +1,265 @@
+/*!
+ * \file   tests/GenericParallelIntegrateTest.cxx
+ * \brief  This test compares the integration of behaviours generated by the
+ * `generic-parallel` interface of `MFront` on batches of integration points
+ * with the integration point by point of the same behaviours generated by the
+ * `generic` interface, on identical states.
+ * \date   09/10/2026
+ * \copyright Use, modification and distribution are subject
+ * to one of the following licences:
+ * - GNU Lesser General Public License (LGPL), Version 3.0. (See accompanying
+ *   file LGPL-3.0.txt)
+ * - CECILL-C,  Version 1.0 (See accompanying files
+ *   CeCILL-C_V1-en.txt and CeCILL-C_V1-fr.txt).
+ */
+
+#include <cmath>
+#include <span>
+#include <string>
+#include <vector>
+#include <cstdlib>
+#include <iostream>
+#include <algorithm>
+#include <string_view>
+#include "MGIS/Behaviour/Behaviour.hxx"
+#include "MGIS/Behaviour/MaterialDataManager.hxx"
+#include "MGIS/Behaviour/Integrate.hxx"
+
+static bool success = true;
+
+static bool check(const bool b, const std::string_view msg) {
+  if (!b) {
+    success = false;
+    std::cerr << msg << '\n';
+  }
+  return b;
+}
+
+//! \return the maximum of the differences relative to the given scale
+static double compare(const std::span<const mgis::real> a,
+                      const std::span<const mgis::real> b,
+                      const double scale) {
+  auto d = 0.;
+  for (std::size_t i = 0; i != a.size(); ++i) {
+    d = std::max(d, std::abs(a[i] - b[i]) / scale);
+  }
+  return d;
+}  // end of compare
+
+//! \brief data managers of the same behaviour, point by point and as a batch
+struct Managers {
+  Managers(const std::string& library,
+           const std::string& gp_library,
+           const std::string& behaviour,
+           const mgis::size_type n)
+      : b(mgis::behaviour::load(
+            library, behaviour, mgis::behaviour::Hypothesis::TRIDIMENSIONAL)),
+        gp_b(
+            mgis::behaviour::load(gp_library,
+                                  behaviour,
+                                  mgis::behaviour::Hypothesis::TRIDIMENSIONAL)),
+        m(b, n),
+        gp_m(gp_b, n) {}
+  //! \brief apply the given function to both managers
+  template <typename Function>
+  void apply(const Function& f) {
+    f(this->m);
+    f(this->gp_m);
+  }
+  /*!
+   * \brief integrate the behaviour with both managers and compare the
+   * results
+   * \param[in] name: name of the test
+   * \param[in] opts: integration options
+   * \param[in] dt: time increment
+   * \param[in] range: range of integration points, all the points if empty
+   */
+  void integrate(const std::string& name,
+                 const mgis::behaviour::BehaviourIntegrationOptions& opts,
+                 const mgis::real dt,
+                 const std::vector<mgis::size_type>& range = {}) {
+    using namespace mgis::behaviour;
+    auto r = BehaviourIntegrationResult{};
+    auto gp_r = BehaviourIntegrationResult{};
+    if (range.empty()) {
+      r = mgis::behaviour::integrate(this->m, opts, dt);
+      // automatic switch to the integration of a batch
+      gp_r = mgis::behaviour::integrate(this->gp_m, opts, dt);
+    } else {
+      r = mgis::behaviour::integrate(this->m, opts, dt, range[0], range[1]);
+      gp_r =
+          mgis::behaviour::integrate(this->gp_m, opts, dt, range[0], range[1]);
+    }
+    check(r.exit_status == gp_r.exit_status, name + ": invalid status");
+    if (r.exit_status == -1) {
+      // the integration point by point stops at the first failure
+      check(r.n == gp_r.n, name + ": invalid index of the first failure");
+      return;
+    }
+    check(std::abs(r.time_step_increase_factor -
+                   gp_r.time_step_increase_factor) < 1e-12,
+          name + ": invalid time step increase factor");
+    constexpr auto eps = 1e-10;
+    check(compare(this->m.s1.thermodynamic_forces,
+                  this->gp_m.s1.thermodynamic_forces, 1e8) < eps,
+          name + ": invalid thermodynamic forces");
+    check(compare(this->m.s1.internal_state_variables,
+                  this->gp_m.s1.internal_state_variables, 1) < eps,
+          name + ": invalid internal state variables");
+    if (opts.integration_type !=
+        IntegrationType::INTEGRATION_NO_TANGENT_OPERATOR) {
+      check(compare(this->m.K, this->gp_m.K, 1e10) < eps,
+            name + ": invalid tangent operators");
+    }
+  }
+  //! \brief behaviour of the generic interface
+  const mgis::behaviour::Behaviour b;
+  //! \brief behaviour of the generic-parallel interface
+  const mgis::behaviour::Behaviour gp_b;
+  //! \brief data manager associated with the behaviour of a point
+  mgis::behaviour::MaterialDataManager m;
+  //! \brief data manager associated with the batch behaviour
+  mgis::behaviour::MaterialDataManager gp_m;
+};
+
+//! \brief viscoplastic behaviour, uniform temperature, several time steps
+static void testNorton(const std::string& library,
+                       const std::string& gp_library) {
+  using namespace mgis::behaviour;
+  constexpr auto n = mgis::size_type{100};
+  auto managers = Managers(library, gp_library, "Norton", n);
+  managers.apply([](MaterialDataManager& m) {
+    setExternalStateVariable(m.s0, "Temperature", 293.15);
+    setExternalStateVariable(m.s1, "Temperature", 293.15);
+  });
+  auto opts = BehaviourIntegrationOptions{};
+  opts.integration_type =
+      IntegrationType::INTEGRATION_CONSISTENT_TANGENT_OPERATOR;
+  for (int step = 0; step != 20; ++step) {
+    managers.apply([](MaterialDataManager& m) {
+      // the strain increment depends on the point
+      for (mgis::size_type i = 0; i != m.n; ++i) {
+        m.s1.gradients[i * m.s1.gradients_stride] +=
+            5e-5 * (1 + static_cast<double>(i) / m.n);
+      }
+    });
+    managers.integrate("Norton, step " + std::to_string(step), opts, 180);
+    managers.apply([](MaterialDataManager& m) { update(m); });
+  }
+}  // end of testNorton
+
+/*!
+ * \brief plastic behaviour, temperature given point by point, integration
+ * of a range of points
+ */
+static void testPlasticity(const std::string& library,
+                           const std::string& gp_library) {
+  using namespace mgis::behaviour;
+  constexpr auto n = mgis::size_type{64};
+  auto managers = Managers(library, gp_library, "Plasticity", n);
+  auto temperatures = std::vector<mgis::real>(n);
+  for (mgis::size_type i = 0; i != n; ++i) {
+    temperatures[i] = 293.15 + 10 * static_cast<double>(i);
+  }
+  managers.apply([&temperatures](MaterialDataManager& m) {
+    setExternalStateVariable(m.s0, "Temperature", temperatures,
+                             MaterialStateManager::EXTERNAL_STORAGE);
+    setExternalStateVariable(m.s1, "Temperature", temperatures,
+                             MaterialStateManager::EXTERNAL_STORAGE);
+    // elastic and plastic points
+    for (mgis::size_type i = 0; i != m.n; ++i) {
+      const auto e = 4e-3 * static_cast<double>(i) / m.n;
+      m.s1.gradients[i * m.s1.gradients_stride] = e;
+      m.s1.gradients[i * m.s1.gradients_stride + 1] = -0.3 * e;
+      m.s1.gradients[i * m.s1.gradients_stride + 2] = -0.3 * e;
+    }
+  });
+  auto opts = BehaviourIntegrationOptions{};
+  opts.integration_type =
+      IntegrationType::INTEGRATION_CONSISTENT_TANGENT_OPERATOR;
+  managers.integrate("Plasticity", opts, 1);
+  managers.integrate("Plasticity, range", opts, 1, {10, 50});
+  opts.integration_type = IntegrationType::PREDICTION_ELASTIC_OPERATOR;
+  managers.integrate("Plasticity, prediction", opts, 1);
+}  // end of testPlasticity
+
+/*!
+ * \brief elastic behaviour: material property given point by point (Young
+ * modulus) and uniform material property (Poisson ratio), integration of a
+ * range of points
+ */
+static void testElasticity(const std::string& library,
+                           const std::string& gp_library) {
+  using namespace mgis::behaviour;
+  constexpr auto n = mgis::size_type{32};
+  auto managers = Managers(library, gp_library, "Elasticity", n);
+  auto young = std::vector<mgis::real>(n);
+  for (mgis::size_type i = 0; i != n; ++i) {
+    young[i] = 150e9 + 1e9 * static_cast<double>(i);
+  }
+  managers.apply([&young](MaterialDataManager& m) {
+    for (auto* const s : {&m.s0, &m.s1}) {
+      setMaterialProperty(*s, "YoungModulus", young,
+                          MaterialStateManager::EXTERNAL_STORAGE);
+      setMaterialProperty(*s, "PoissonRatio", 0.3);
+      setExternalStateVariable(*s, "Temperature", 293.15);
+    }
+    for (mgis::size_type i = 0; i != m.n; ++i) {
+      m.s1.gradients[i * m.s1.gradients_stride] = 1e-3;
+    }
+  });
+  auto opts = BehaviourIntegrationOptions{};
+  opts.integration_type =
+      IntegrationType::INTEGRATION_CONSISTENT_TANGENT_OPERATOR;
+  managers.integrate("Elasticity", opts, 1);
+  managers.apply([](MaterialDataManager& m) {
+    for (mgis::size_type i = 0; i != m.n; ++i) {
+      m.s1.gradients[i * m.s1.gradients_stride] = 2e-3;
+    }
+  });
+  managers.integrate("Elasticity, range", opts, 1, {5, 20});
+}  // end of testElasticity
+
+//! \brief failure at all points
+static void testFailure(const std::string& library,
+                        const std::string& gp_library) {
+  using namespace mgis::behaviour;
+  auto managers = Managers(library, gp_library, "NortonFailure", 10);
+  managers.apply([](MaterialDataManager& m) {
+    setExternalStateVariable(m.s0, "Temperature", 293.15);
+    setExternalStateVariable(m.s1, "Temperature", 293.15);
+  });
+  auto opts = BehaviourIntegrationOptions{};
+  managers.integrate("NortonFailure", opts, 1);
+  managers.integrate("NortonFailure, range", opts, 1, {3, 7});
+  // the debugging functions are not available for batches
+  auto& gp_m = managers.gp_m;
+  check(
+      [&gp_m, &opts] {
+        try {
+          static_cast<void>(integrate_debug(gp_m, opts, 1));
+        } catch (...) {
+          return true;
+        }
+        return false;
+      }(),
+      "integrate_debug must be refused for batch behaviours");
+}  // end of testFailure
+
+int main(const int argc, const char* const* const argv) {
+  if (argc != 3) {
+    std::cerr << "usage: " << argv[0]
+              << " generic-parallel-library generic-library\n";
+    return EXIT_FAILURE;
+  }
+  try {
+    testNorton(argv[2], argv[1]);
+    testPlasticity(argv[2], argv[1]);
+    testElasticity(argv[2], argv[1]);
+    testFailure(argv[2], argv[1]);
+  } catch (std::exception& e) {
+    std::cerr << e.what() << '\n';
+    return EXIT_FAILURE;
+  }
+  return success ? EXIT_SUCCESS : EXIT_FAILURE;
+}  // end of main
-- 
2.43.0

